File: smoke_tests/csharp/open_redirect.cs
// Open Redirect vulnerabilities in C#
using System;
using System.Web;
using System.Web.Mvc;

namespace VulnerableApp
{
    public class RedirectController : Controller
    {
        // Test 1: Direct redirect from query parameter
        public ActionResult Login()
        {
            string returnUrl = Request.QueryString["returnUrl"];
            // VULNERABLE: Unvalidated redirect
            return Redirect(returnUrl);
        }

        // Test 2: RedirectToAction with external URL
        public ActionResult Logout()
        {
            string target = Request.QueryString["target"];
            // VULNERABLE: Can redirect to external URL
            return Redirect(target);
        }

        // Test 3: Response.Redirect
        public void DirectRedirect()
        {
            string url = Request.QueryString["url"];
            // VULNERABLE: Unvalidated redirect
            Response.Redirect(url);
        }

        // Test 4: Partial validation bypass
        public ActionResult SafeRedirect()
        {
            string url = Request.QueryString["url"];
            // VULNERABLE: Can be bypassed with //evil.com
            if (url.StartsWith("/"))
            {
                return Redirect(url);
            }
            return RedirectToAction("Index");
        }

        // Test 5: URL in form data
        public ActionResult ProcessForm()
        {
            string nextPage = Request.Form["next"];
            // VULNERABLE: Form data redirect
            return Redirect(nextPage);
        }

        // Test 6: Cookie-based redirect
        public ActionResult RestoreSession()
        {
            string returnUrl = Request.Cookies["returnUrl"]?.Value;
            // VULNERABLE: Cookie can be manipulated
            if (!string.IsNullOrEmpty(returnUrl))
            {
                return Redirect(returnUrl);
            }
            return RedirectToAction("Index");
        }

        // Test 7: Header-based redirect
        public ActionResult HandleRequest()
        {
            string referer = Request.Headers["X-Return-Url"];
            // VULNERABLE: Header-based redirect
            if (!string.IsNullOrEmpty(referer))
            {
                return Redirect(referer);
            }
            return View();
        }

        // Test 8: JavaScript redirect
        public ActionResult JsRedirect()
        {
            string url = Request.QueryString["url"];
            // VULNERABLE: JavaScript redirect with user URL
            return Content($"<script>window.location='{url}';</script>", "text/html");
        }

        // Test 9: Meta refresh redirect
        public ActionResult MetaRedirect()
        {
            string target = Request.QueryString["target"];
            // VULNERABLE: Meta refresh with user URL
            return Content($"<meta http-equiv='refresh' content='0;url={target}'>", "text/html");
        }

        // Test 10: Domain validation bypass
        public ActionResult DomainCheck()
        {
            string url = Request.QueryString["url"];
            // VULNERABLE: evil.example.com contains example.com
            if (url.Contains("example.com"))
            {
                return Redirect(url);
            }
            return RedirectToAction("Index");
        }

        // Test 11: LocalRedirect misuse
        public ActionResult AfterAuth()
        {
            string returnUrl = Request.QueryString["returnUrl"];
            // VULNERABLE: If Url.IsLocalUrl is not called first
            if (!string.IsNullOrEmpty(returnUrl))
            {
                return Redirect(returnUrl);
            }
            return RedirectToAction("Index");
        }
    }
}

File: smoke_tests/cpp/open_redirect.cpp
// Open Redirect vulnerabilities in C++ (CGI/Web context)
#include <iostream>
#include <string>
#include <cstdlib>

// Test 1: Direct redirect from query parameter
void handle_redirect() {
    char* url = getenv("QUERY_STRING");
    // VULNERABLE: Unvalidated redirect
    std::cout << "Status: 302 Found\r\n";
    std::cout << "Location: " << url << "\r\n\r\n";
}

// Test 2: Redirect from POST data
void login_redirect(const std::string& return_url) {
    // VULNERABLE: return_url not validated
    std::cout << "Status: 302 Found\r\n";
    std::cout << "Location: " << return_url << "\r\n\r\n";
}

// Test 3: Partial validation bypass
void redirect_checked(const std::string& url) {
    // VULNERABLE: Can be bypassed with //evil.com or javascript:
    if (url[0] == '/') {
        std::cout << "Status: 302 Found\r\n";
        std::cout << "Location: " << url << "\r\n\r\n";
    }
}

// Test 4: Meta refresh redirect
void meta_redirect(const std::string& target) {
    std::cout << "Content-Type: text/html\r\n\r\n";
    // VULNERABLE: Unvalidated target URL
    std::cout << "<html><head><meta http-equiv='refresh' content='0;url=" << target << "'></head></html>";
}

// Test 5: JavaScript redirect
void js_redirect(const std::string& url) {
    std::cout << "Content-Type: text/html\r\n\r\n";
    // VULNERABLE: User-controlled URL in JavaScript
    std::cout << "<script>window.location='" << url << "';</script>";
}

// Test 6: Header injection through redirect
void redirect_with_header(const std::string& target) {
    // VULNERABLE: target can contain CRLF to inject headers
    std::cout << "Status: 302 Found\r\n";
    std::cout << "Location: " << target << "\r\n";
    std::cout << "Content-Type: text/html\r\n\r\n";
}

// Test 7: Subdomain-based redirect bypass
void redirect_same_domain(const std::string& url) {
    // VULNERABLE: evil.example.com matches *.example.com
    if (url.find("example.com") != std::string::npos) {
        std::cout << "Status: 302 Found\r\n";
        std::cout << "Location: " << url << "\r\n\r\n";
    }
}

// Test 8: Encoded redirect bypass
void redirect_decoded(const std::string& encoded_url) {
    // VULNERABLE: URL decoding may reveal malicious URL
    std::string decoded = encoded_url;  // Assume decoded
    std::cout << "Status: 302 Found\r\n";
    std::cout << "Location: " << decoded << "\r\n\r\n";
}
